src/services/geometry/richgeo/IrtGeo.cc

```diff
@@ -1,51 +1,83 @@
-// Copyright 2022, Christopher Dilks
+// Copyright (C) 2022, 2023, Christopher Dilks
 // Subject to the terms in the LICENSE file found in the top-level directory.
-//
-//
 
 #include "IrtGeo.h"
 
 // constructor: creates IRT-DD4hep bindings using main `Detector` handle `*det_`
-richgeo::IrtGeo::IrtGeo(std::string detName_, dd4hep::Detector *det_, bool verbose_) :
-  m_detName(detName_), m_det(det_), m_log(Logger::Instance(verbose_))
+richgeo::IrtGeo::IrtGeo(std::string detName_, dd4hep::Detector *det_, std::shared_ptr<spdlog::logger> log_) :
+  m_detName(detName_), m_det(det_), m_log(log_)
 {
   Bind();
 }
 
 // alternate constructor: use compact file for DD4hep geometry (backward compatibility)
-richgeo::IrtGeo::IrtGeo(std::string detName_, std::string compactFile_, bool verbose_) :
-  m_detName(detName_), m_log(Logger::Instance(verbose_))
+richgeo::IrtGeo::IrtGeo(std::string detName_, std::string compactFile_, std::shared_ptr<spdlog::logger> log_) :
+  m_detName(detName_), m_log(log_)
 {
-  // compact file name; if it's not been specified, try to find the default one
-  std::string compactFile;
-  if(compactFile_=="") {
-    std::string DETECTOR_PATH(getenv("DETECTOR_PATH"));
-    std::string DETECTOR_CONFIG(getenv("DETECTOR_CONFIG"));
-    if(DETECTOR_PATH.empty() || DETECTOR_CONFIG.empty()) {
-      m_log.PrintError("cannot find default compact file, since env vars DETECTOR_PATH and DETECTOR_CONFIG are not set");
-      return;
-    }
-    compactFile = DETECTOR_PATH + "/" + DETECTOR_CONFIG + ".xml";
-  } else compactFile = compactFile_;
-
   // build DD4hep detector from compact file
   m_det = &dd4hep::Detector::getInstance();
-  m_det->fromXML(compactFile);
+  m_det->fromXML(compactFile_);
 
   // set IRT and DD4hep geometry handles
   Bind();
 }
 
+// Bind() -----------------------------------------
 // set IRT and DD4hep geometry handles
 void richgeo::IrtGeo::Bind() {
   // DD4hep geometry handles
   m_detRich = m_det->detector(m_detName);
   m_posRich = m_detRich.placement().position();
+
   // IRT geometry handles
   m_irtDetectorCollection = new CherenkovDetectorCollection();
   m_irtDetector = m_irtDetectorCollection->AddNewDetector(m_detName.c_str());
+
+  // cellID conversion
+  m_cellid_converter = std::make_shared<const dd4hep::rec::CellIDPositionConverter>(*m_det);
+
+  // define the `cell ID -> pixel position` converter, correcting the returned
+  // `Position` to be at the sensor surface
+  m_irtDetector->m_ReadoutIDToPosition = [this] (auto cell_id) {
```

Comment:
Same here: `this` as `const` capture? I'm worried about potential for unintentional thread unsafe write when using non const operations on some things that come in through the captures.

Reply:
I couldn't find an example of `const` capture, could you clarify what you mean?

I'm wondering if instead I could capture `*this` (a copy of `this`), but the way it's currently written I don't think that will work since this lambda is defined before members such as `m_irtDetector` are completely built.

I can move the definition of this lambda to somewhere after the `DD4hep_to_IRT` (which sets everything that's needed in the capture), and then capturing `*this` might be okay, since that's the version of `this` I want to copy. My thought is that if `[*this]` indeed makes a copy, then it doesn't matter what happens to the original `this`. What do you think?

An alternative approach is a giant `std::unordered_map`... but I think that would be slower since this lambda already uses the CellID-to-position converter, and the number of hits is << the number of pixels.

Reply:
Capturing `*this` leads to a double free error, and requires `IrtGeo` to have no pure virtual functions.

Instead, I propose capturing only the members I need by value (except for the logger, captured by reference). I also moved the definition of the lambda to _after_ `DD4hep_to_IRT`, so that everything it captures is fully defined.

```cpp
  m_irtDetector->m_ReadoutIDToPosition = [
    &m_log = this->m_log, // capture logger by reference
    // capture instance members by value, so those owned by `this` are not mutable here
    cell_mask        = this->m_irtDetector->GetReadoutCellMask(),
    cellid_converter = this->m_cellid_converter,
    sensor_info      = this->m_sensor_info
  ] (auto cell_id) {
    // ...
  };
```

Reply:
changes: 446c18df83163df237e7aae15202cc8a67e476ff

Reply:
tested in https://github.com/eic/EICrecon/pull/393 and works.